File: core/runtime/semantic_lock_engine.py
from __future__ import annotations

from threading import Lock

from typing import Dict


LOCKS: Dict[str, Lock] = {}


def acquire_semantic_lock(
    key: str,
) -> bool:

    if key not in LOCKS:

        LOCKS[key] = Lock()

    return LOCKS[key].acquire(
        blocking=False,
    )


def release_semantic_lock(
    key: str,
) -> None:

    if key in LOCKS:

        try:
            LOCKS[key].release()

        except RuntimeError:
            pass

File: src/runtime/semanticLockEngine.ts
/**
 * Converted from Python: core/runtime/semantic_lock_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "./pyCompat.js";
// from threading import ... (unmapped)

var LOCKS: Record<string, any> = {};
export function acquireSemanticLock(key: any): any {
  if (!py.contains(LOCKS, key)) {
    py.setItem(LOCKS, key, py.lock());
  }
  return py.at(LOCKS, key).acquire(false);
}
export function releaseSemanticLock(key: any): any {
  if (py.contains(LOCKS, key)) {
    try {
      py.at(LOCKS, key).release();
    } catch (_e: any) {
    }
  }
}
